### ISSUE TITLE
cAdvisor leaking journalctl processes

### ISSUE THREAD

**Issue by vishh** [kubernetes/kubernetes#34965]
Extracted from https://github.com/kubernetes/kubernetes/pull/23491#issuecomment-251777767

```
This appears to be leaking journalctl processes originating here: https://github.com/kubernetes/kubernetes/blob/master/vendor/github.com/google/cadvisor/utils/oomparser/oomparser.go#L169
```

https://github.com/kubernetes/kubernetes/pull/23491 set kubelet unit files to restart just the kubelet process and this ends up orphaning and leaking `journalctl` processes launched by cAdvisor.
AFAIK, this behavior can leak many other processes like `du`, `ls`, `mount`, etc. 

@kubernetes/sig-node we need to fix this. 

I recommend reverting https://github.com/kubernetes/kubernetes/pull/23491 to begin with as suggested by @wwwtyro

**Comment by vishh** [comment]
Assigning to @dchen1107 to find an appropriate owner.

**Comment by vishh** [comment]
cc @saad-ali @rootfs we need to come up with a better solution for https://github.com/kubernetes/kubernetes/pull/23491#issuecomment-205556234

**Comment by vishh** [comment]
Reverting #23491 might cause storage issues. So let's find out a better solution.

**Comment by AdoHe** [comment]
@vishh I did something on node-problem-detector to use go-systemd instead of journalctl. You may want to take a look [here](https://github.com/kubernetes/node-problem-detector/pull/33)

**Comment by onorua** [comment]
what about initial proposal to move volumes to FUSE in pod scope?

**Comment by dchen1107** [comment]
@AdoHe Thanks for the suggestion of using go-systemd instead of directly invoke journalctl in cAdvisor to parse kernel's oom event. That is less error prone in my opinion. We are going to get your pr merged for 1.5 release. Thanks!

Back to this particular issue, we should retire this piece of code in cAdvisor eventually once node-problem-detector is deployed to most clusters. If we couldn't retire that code, we are going to refactor that code as what you suggested. 

But that doesn't solve all the issues reported here even `journalctl -k -f`  is the only long running children process today. `du`, `ls`, `mount` are not. But the current model of setting KillMode for kubelet to process is problematic to fix glusterfs umount issue. @onorua, rootfs@: Any concern to set that to control-group / mixed? If you guys ok with that, I can send a pr fix that shortly.

**Comment by euank** [comment]
I think setting `KillMode=control-group` is a good workaround for now.

I think that both node-problem-detector and cadvisor (if it keeps doing oom stuff) shouldn't be using go-systemd or journald. Rather, parsing `/dev/kmsg` seems like the right way to do this.

To parse `/dev/kmsg` and get times though, we'll want a syslog format parser, so it adds some complexity there, but at least it'll let us do exactly one thing for every linux, rather than 5 different things

**Comment by dims** [comment]
This needs to be triaged as a release-blocker or not for 1.5 @vishh @euank @AdoHe @dchen1107

**Comment by euank** [comment]
We should use kmsg instead of journald / sd-journal since cadvisor *only* cares about oom messages IMO.

Duplicating my work from https://github.com/kubernetes/node-problem-detector/pull/41 against cadvisor should fix this, and I've been planning to do that regardless.

**Comment by dims** [comment]
@euank ack thanks for the detail. Does not sound like something we need for 1.5, moving to 1.6. please switch it right back if i misunderstood.

**Comment by euank** [comment]
This one seems a bit worrisome as bugs go. The fix I suggest above is a bit drastic of a change for post-freeze, but leaking processes isn't great.

If we think this should be part of 1.5 (cc @vishh) I can definitely prioritize getting my fix into cadvisor monday or tuesday

**Comment by jonboulle** [comment]
Is this missing 1.5? It's pretty bad.

**Comment by ethernetdan** [comment]
@vishh @dchen1107 can this be closed?

**Comment by euank** [comment]
@ethernetdan the bug still exists. `KillMode=process` is still set in a few places (including cluster/saltbase)  still too.

**Comment by dchen1107** [comment]
Yes, the bug still exists. The original proposal I made to set `KillMode=control-group` was rejected by storage folks: https://github.com/kubernetes/kubernetes/pull/35479#issuecomment-256551084

On another hand, the initial reason we introduced oomparser to cAdvisor is for detecting the oom events. The current situation is that 
1) docker already properly report container's memory-exceeded oom event long time ago.
2) sys oom was spotted and reported by NPD too. 

Beyond above two use cases, currently there is no other use for this information from cAdvisor. Like what I suggested at https://github.com/kubernetes/kubernetes/issues/34965#issuecomment-255182160, we should completely retire this from cAdvisor.

If we plan to still use the information from cAdvisor, I think @euank 's pr #1544 should be the proper fix.

**Comment by dchen1107** [comment]
Move this one to 1.7

**Comment by ingvagabund** [comment]
@dchen1107 this seems as a tough problem. Have there been any progress since March?

**Comment by euank** [comment]
I do think the PR I have against cadvisor (https://github.com/google/cadvisor/pull/1544) is still a good solution to this problem.

**Comment by jsafrane** [comment]
Summing up the discussion here:

* kubelet service with `KillMode=process` does not kill cAdvisor's probes.
* kubelet service with `KillMode=control-group` kills fuse daemons and thus unmounts gluster volumes from running pods.

I have a PR #48430 where kubelet running in a container runs /bin/mount on the host, outside of kubelet's container cgroup using systemd-run. I will try the same approach on the non-containerizes kubelet. This way, all fuse daemons would run in another systemd service (and cgroup) and kubelet could run with `KillMode=control-group`.

**Comment by jsafrane** [comment]
filled #49640 to run *all* fuse daemons in their own systemd scope (=cgroup) so they won't be killed by `KillMode=control-group`

**Comment by mtaufen** [comment]
I would really like to get this journalctl leakage fixed wrt #1725. I like the idea of using kmsg instead of journald, but I think the simplest quick fix might just be to set the Pdeathsig to SIGKILL on the journalctl command's SysProcAttr. Can we agree on something simple like that so we at least fix the leak while we figure out an ideal solution?

**Comment by euank** [comment]
Note, the cadvisor change has to be vendored back over too, so this close might be premature (thanks github 'fix' keywod)
